controllers/clustertopology_controller_compute.go

```diff
@@ -66,10 +73,116 @@ type machineDeploymentTopologyState struct {
 	infrastructureMachineTemplate *unstructured.Unstructured   //nolint:structcheck
 }
 
-// Gets the ClusterClass and the referenced templates to be used for a managed Cluster topology.
-func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clusterv1.Cluster) (*clusterTopologyClass, error) {
-	// TODO: add get class logic; also remove nolint exception from clusterTopologyClass and machineDeploymentTopologyClass
-	return nil, nil
+// getClass gets the ClusterClass and the referenced templates to be used for a managed Cluster topology. It also converts
+// and patches all ObjectReferences in ClusterClass and ControlPlane to the latest apiVersion of the current contract.
+// NOTE: This function assumes that cluster.Spec.Topology.Class is set.
+func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clusterv1.Cluster) (_ *clusterTopologyClass, reterr error) {
+	class := &clusterTopologyClass{
+		clusterClass:       &clusterv1.ClusterClass{},
+		machineDeployments: map[string]machineDeploymentTopologyClass{},
+	}
+
+	// get ClusterClass
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: cluster.Spec.Topology.Class, Namespace: cluster.Namespace}, class.clusterClass); err != nil {
+		return nil, errors.Wrapf(err, "failed to retrieve ClusterClass %q in namespace %q", cluster.Spec.Topology.Class, cluster.Namespace)
+	}
+
+	// We use the patchHelper to patch potential changes to the ObjectReferences in ClusterClass.
+	patchHelper, err := patch.NewHelper(class.clusterClass, r.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, class.clusterClass); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch ClusterClass %q in namespace %q", class.clusterClass.Name, class.clusterClass.Namespace)})
+		}
+	}()
+
+	// get ClusterClass.spec.infrastructure
+	if class.infrastructureClusterTemplate, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.Infrastructure.Ref); err != nil {
+		return nil, err
+	}
+
+	// get ClusterClass.spec.controlPlane
+	if class.controlPlane.template, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.ControlPlane.Ref); err != nil {
+		return nil, err
+	}
+
+	if class.clusterClass.Spec.ControlPlane.MachineInfrastructure != nil && class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil {
+		// get ClusterClass.spec.controlPlane.machineInfrastructure
+		if class.controlPlane.infrastructureMachineTemplate, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref); err != nil {
+			return nil, err
+		}
+	}
+
+	// if ControlPlaneTemplate.spec.template.spec.machineTemplate.infrastructureRef is set, convert and patch it
+	// returns an error if the ObjectReference is partially set
```

Comment:
After thinking a little bit more about this, I'm getting of the idea that upgrading references in the ControlPlaneTemplate object should not a CAPI responsibility, but instead the single control place providers (e.g. KCP) should be responsible to keep their own objects up to date.

This will make the contract between CAPI and the provider simpler:
e.g. in order to allow this change it is required to that the validation web hooks for the ControlPlaneTemplate allows for this change, and this is not currently defined in the proposed amendment in the contract. 
By handling over the update responsibly to the provider, the additional change won't be required. 

Last but not least, working with unstructured is really ugly and error prone.

The only downside of this, is that this requires a new controller in KCP (and potentially in all the other control plane controllers using machines)

@vincepri @sbueringer  opinions?

Reply:
I think a clear separation of concern should be preferred here. I think the only reason why we considered doing this in the ClusterClass reconciler is "because there is currently nobody else who takes care of this". While we also have the KubeadmControlPlane controller "under our control" this feels even stranger (to me) when thinking about updating references in 3rd part ControlPlanes.

So overall I think we should keep it consistent with the other places where where we do this (afaik).

Reply:
I'm +1 to remove from this PR and create a new controller in KCP

Reply:
+1 to remove it from this PR

Reply:
Done

Reply:
Started a discussion about follow-up tasks in Slack: https://kubernetes.slack.com/archives/C8TSNPY4T/p1627460469457000 (which then might lead to a new issue)